src/EconomicStrategy/EconomicStrategyManager.ts

```diff
@@ -182,59 +196,102 @@ export class EconomicStrategyManager {
-      }
+      minBalance = minBalance.add(costOfExecutingFutureTransactions);
     }
 
-    if (minBalance) {
-      return currentBalance.gt(minBalance);
-    }
-    return true;
+    return currentBalance.gt(minBalance);
   }
 
-  private async isProfitable(txRequest: ITxRequest): Promise<boolean> {
+  private async isClaimingProfitable(txRequest: ITxRequest, gasPrice: BigNumber): Promise<boolean> {
     const paymentModifier = await txRequest.claimPaymentModifier();
-    const claimingGas = new BigNumber(CLAIMING_GAS_ESTIMATE);
-    // NOTE - Here we use the legacy (web3 version) of getGasPrice instead of the
-    // networkGasPrice() call. This is to ensure backwards compatibility with older
-    // versions of the TimeNode. If TimeNodes disagree about the current gasPrice,
-    // we notice disambiguous behavior between client versions which make both versions suck more.
-    // Instead we opt for maintaining the sucky version for now, to keep the levels of
-    // suck constant instead of increasing them.
-    const claimingGasPrice = await this.util.getGasPrice();
-    const claimingGasCost = claimingGasPrice.times(claimingGas);
-
+    const claimingGasCost = gasPrice.times(CLAIMING_GAS_ESTIMATE);
     const reward = txRequest.bounty.times(paymentModifier).minus(claimingGasCost);
+    const minProfitability = this.strategy.minProfitability || new BigNumber(0);
+
+    return reward.gte(minProfitability);
+  }
+
+  private normalizeWaitTimes = (temporalUnit: number, stats: EthGasStationInfo) => {
+    let normalizedTimes = {
+      safeLow: stats.safeLowWait.mul(10),
+      avg: stats.avgWait.mul(10),
+      fast: stats.fastWait.mul(10),
+      fastest: stats.fastestWait.mul(10)
+    };
+
+    if (temporalUnit === 1) {
+      // Normalize the estimate
+      const { blockTime } = stats;
+      normalizedTimes = {
+        safeLow: normalizedTimes.safeLow.div(blockTime).round(),
+        avg: normalizedTimes.avg.div(blockTime).round(),
+        fast: normalizedTimes.fast.div(blockTime).round(),
+        fastest: normalizedTimes.fastest.div(blockTime).round()
+      };
+    }
 
-    const minProfitability = this.strategy.minProfitability;
+    return normalizedTimes;
+  };
 
-    if (minProfitability && minProfitability.gt(0)) {
-      return reward.gt(minProfitability);
+  // tslint:disable-next-line:cognitive-complexity
+  private async smartGasEstimation(txRequest: ITxRequest): Promise<BigNumber | null> {
+    const gasStats = await W3Util.getEthGasStationStats();
+    if (!gasStats) {
+      return null;
     }
 
-    return true;
+    const { temporalUnit } = txRequest;
+    const normTimes = this.normalizeWaitTimes(temporalUnit, gasStats);
+    // Reserved, need to send transaction before it goes to general window.
+    if (await txRequest.inReservedWindow()) {
+      let timeLeftInReservedWindow = (await txRequest.now()).sub(txRequest.reservedWindowEnd);
```

Comment:
Lines 263  - 277 and 280 - 294, look super similar. Can they be extracted to a function?